### DISCUSSION TITLE
BacklinksProcessor running documents.update: API timeout

### DISCUSSION THREAD

**Discussion by lhqs1314** [outline/outline#4653] [Self Hosting]
Outline: 
Install: From source
Version: v0.66.2

Desktop :
OS: MacOS 12.5
Browser: Chrome

Description: 
It works fine after starting with "yarn start", but after about 2 minutes the interface gets stuck and the logs are paused at "{"name": "documents.update", "label": "worker"," level": "info", "message": "BacklinksProcessor running documents.update"}", the document cannot be requested normally, and there are no other error messages

![image](https://user-images.githubusercontent.com/29948603/210962672-c9bf006d-7bb1-4608-a217-5b09445b8fc7.png)

Log information: 
`{"label":"http","level":"info","message":"  \u001b[90m<--\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=3d1dd43e-89a8-4157-9aa8-215613da592d\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m<--\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=3348f30e-8670-411f-af5b-89dd1bb27b77\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m-->\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=3d1dd43e-89a8-4157-9aa8-215613da592d\u001b[39m \u001b[36m302\u001b[39m \u001b[90m75ms\u001b[39m \u001b[90m977b\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m<--\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=5d6eac46-196e-4a9a-b9e2-71bf3ef02242\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m-->\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=3348f30e-8670-411f-af5b-89dd1bb27b77\u001b[39m \u001b[36m302\u001b[39m \u001b[90m96ms\u001b[39m \u001b[90m977b\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m-->\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=5d6eac46-196e-4a9a-b9e2-71bf3ef02242\u001b[39m \u001b[36m302\u001b[39m \u001b[90m98ms\u001b[39m \u001b[90m977b\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m<--\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=a1b9563e-d1f6-4ae9-a174-09c817a00890\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m<--\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=753370f7-9d14-42e0-a997-faf177c00537\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m-->\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=a1b9563e-d1f6-4ae9-a174-09c817a00890\u001b[39m \u001b[36m302\u001b[39m \u001b[90m54ms\u001b[39m \u001b[90m977b\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m-->\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=753370f7-9d14-42e0-a997-faf177c00537\u001b[39m \u001b[36m302\u001b[39m \u001b[90m67ms\u001b[39m \u001b[90m977b\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m<--\u001b[39m \u001b[1mPOST\u001b[22m \u001b[90m/api/views.create\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m-->\u001b[39m \u001b[1mPOST\u001b[22m \u001b[90m/api/views.create\u001b[39m \u001b[32m200\u001b[39m \u001b[90m117ms\u001b[39m \u001b[90m632b\u001b[39m"}
{"name":"views.create","modelId":"99f3049e-9896-41a9-89e9-856f42d458f3","attempt":0,"label":"worker","level":"info","message":"Processing views.create"}
{"name":"views.create","modelId":"99f3049e-9896-41a9-89e9-856f42d458f3","label":"worker","level":"info","message":"WebhookProcessor running views.create"}
{"label":"http","level":"info","message":"  \u001b[90m<--\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=bb6e1f7c-c7e3-42f5-ab33-af5a616e6c93\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m-->\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=bb6e1f7c-c7e3-42f5-ab33-af5a616e6c93\u001b[39m \u001b[36m302\u001b[39m \u001b[90m38ms\u001b[39m \u001b[90m977b\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m<--\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=3d1dd43e-89a8-4157-9aa8-215613da592d\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m-->\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=3d1dd43e-89a8-4157-9aa8-215613da592d\u001b[39m \u001b[36m302\u001b[39m \u001b[90m62ms\u001b[39m \u001b[90m977b\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m<--\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=a1b9563e-d1f6-4ae9-a174-09c817a00890\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m<--\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=5d6eac46-196e-4a9a-b9e2-71bf3ef02242\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m-->\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=a1b9563e-d1f6-4ae9-a174-09c817a00890\u001b[39m \u001b[36m302\u001b[39m \u001b[90m69ms\u001b[39m \u001b[90m977b\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m<--\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=3348f30e-8670-411f-af5b-89dd1bb27b77\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m-->\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=5d6eac46-196e-4a9a-b9e2-71bf3ef02242\u001b[39m \u001b[36m302\u001b[39m \u001b[90m77ms\u001b[39m \u001b[90m977b\u001b[39m"}
{"label":"http","level":"info","message":"  \u001b[90m-->\u001b[39m \u001b[1mGET\u001b[22m \u001b[90m/api/attachments.redirect?id=3348f30e-8670-411f-af5b-89dd1bb27b77\u001b[39m \u001b[36m302\u001b[39m \u001b[90m64ms\u001b[39m \u001b[90m977b\u001b[39m"}
{"label":"multiplayer","level":"info","message":"Persisting 78af5836-9325-4130-97ba-3a9ca52ec51b, attributed to c0206760-63a5-48e8-9325-8f4aa9f0f59b"}
{"name":"documents.update","attempt":0,"label":"worker","level":"info","message":"Processing documents.update"}
{"name":"documents.update","label":"worker","level":"info","message":"BacklinksProcessor running documents.update"}`

All log
[outline.log](https://github.com/outline/outline/files/10358834/outline.log)

One additional information: 
"/home/ec2-user/outline/build/server/index.js" This process is using close to 100% of the CPU
![image](https://user-images.githubusercontent.com/29948603/211030080-55f9e080-d509-462b-bf64-3d893c7716e7.png)

How do I locate the cause and fix it, please? 🙏

**Comment by tommoor** [comment]
Never seen this issue before
Does it always stop responding on the `BacklinksProcessor`?
Can you still SSH into the box?

**Reply by lhqs1314** [reply]
Yes    
![image](https://user-images.githubusercontent.com/29948603/211028941-3b700e88-00a3-4901-baeb-8b15db34a5f9.png)

**Reply by lhqs1314** [reply]
One additional information
"/home/ec2-user/outline/build/server/index.js" This process is using close to 100% of the CPU  
![image](https://user-images.githubusercontent.com/29948603/211030411-22e44940-2c74-4eb6-be36-ce02ddd06af1.png)

**Reply by tommoor** [reply]
Sounds like an infinite loop, though I'm not sure how that would happen currently. What does the content of the document look like?

**Reply by lhqs1314** [reply]
Currently just starting to use, not adding too much content
![image](https://user-images.githubusercontent.com/29948603/211480980-6f2e1947-0f96-43f3-b6ee-46eb0625ba4c.png)

**Reply by tommoor** [reply]
If you can export as Markdown and send the page content to hello@getoutline.com it might offer some clues, thanks.

**Comment by murilopolese** [comment]
Hey!!

I have a similar issue running from a docker container. All the requests to the service hang on pending so I can't access the document to see if there is a problem.

I'm curious, what does this log about `BacklinksProcessor` means, @tommoor? 
Would there be a way to try to turn this process off or start without it?

**Reply by tommoor** [reply]
> Would there be a way to try to turn this process off or start without it?

No, it's an integral part of the platform.

I'm not sure we ever received the information required from the previous poster to debug this – so if you have the document content please send it through.

**Reply by murilopolese** [reply]
I'm still stuck outside so I can't give you the document. Also it contains some personal information from a student and it would be weird to send that away :x

I see on the `dev` script there is a flag where you can say what services to spawn:

```
"NODE_ENV=development yarn concurrently -n api,collaboration  -c \"blue,magenta\" \"node --inspect=0.0.0.0 build/server/index.js --services=collaboration,websockets,admin,web,worker\"",
```

I was wondering if there is a way to not pass the `worker` service just so I can backup my data and come back to this issue on a less pressing issue. Do you think this could be a strangled resource? I can see the CPU usage is peaking...

**Reply by tommoor** [reply]
You could turn off the `worker` process to load the application, however the worker process is required to create a backup. Without the content of the document in question it's going to be hard to debug further

**Reply by murilopolese** [reply]
> Without the content of the document in question it's going to be hard to debug further

What would you look at on this file?

> No, it's an integral part of the platform.
> You could turn off the worker process to load the application, however the worker process is required to create a backup.

The worker process apparently does this `BacklinksProcessor` task as well, right? Is that related to backup?
I'm wondering if this issue has something to do with the box I'm running the application not having enough resources to run this task.
